runners/direct-java/src/main/java/org/apache/beam/runners/direct/AggregatorContainer.java

```diff
@@ -0,0 +1,154 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.direct;
+
+import org.apache.beam.sdk.transforms.Aggregator;
+import org.apache.beam.sdk.transforms.Aggregator.AggregatorFactory;
+import org.apache.beam.sdk.transforms.Combine.CombineFn;
+import com.google.common.base.Preconditions;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * AccumT container for the current values associated with {@link Aggregator Aggregators}.
+ */
+public class AggregatorContainer {
+
+  private static class AggregatorInfo<InputT, AccumT, OutputT>
+      implements Aggregator<InputT, OutputT> {
+    private final String name;
+    private final CombineFn<InputT, AccumT, OutputT> combiner;
+    private AccumT accumulator = null;
+    private boolean committed = false;
+
+    private AggregatorInfo(String name, CombineFn<InputT, AccumT, OutputT> combiner) {
+      this.name = name;
+      this.combiner = combiner;
+    }
+
+    @Override
+    public void addValue(InputT input) {
+      Preconditions.checkState(!committed, "Cannot addValue after committing");
+      if (accumulator == null) {
+        accumulator = combiner.createAccumulator();
+      }
+      accumulator = combiner.addInput(accumulator, input);
+    }
+
+    public OutputT getOutput() {
+      return accumulator == null ? null : combiner.extractOutput(accumulator);
+    }
+
+    private void merge(AggregatorInfo<?, ?, ?> other) {
+      // TODO: Make sure the CombineFn and accumulators are compatible.
+      AggregatorInfo<InputT, AccumT, OutputT> otherSafe =
+          (AggregatorInfo<InputT, AccumT, OutputT>) other;
+      mergeSafe(otherSafe);
+    }
+
+    private void mergeSafe(AggregatorInfo<InputT, AccumT, OutputT> other) {
+      if (accumulator == null) {
+        accumulator = other.accumulator;
+      } else if (other.accumulator != null) {
+        accumulator = combiner.mergeAccumulators(Arrays.asList(accumulator, other.accumulator));
+      }
+    }
+
+    @Override
+    public String getName() {
+      return name;
+    }
+
+    @Override
+    public CombineFn<InputT, ?, OutputT> getCombineFn() {
+      return combiner;
+    }
+  }
+
+  private final HashMap<String, AggregatorInfo<?, ?, ?>> accumulators = new HashMap<>();
```

Comment:
If this is a `ConcurrentMap`, you can merge without requiring the lock by using `replace(key, oldAccumulator, newAccumulator)`

This is also the way the direct runner generally does concurrency-safe operations